integration/alertmanager_test.go

```diff
@@ -575,3 +575,196 @@ func TestAlertmanagerSharding(t *testing.T) {
+				instances = append(instances, am)
+				return am
+			}
+
+			// Helper to stop the most recently started instance.
+			popInstance := func() {
+				require.Greater(t, len(instances), 0)
+				last := len(instances) - 1
+				require.NoError(t, s.Stop(instances[last]))
+				instances = instances[:last]
+			}
+
+			// Helper to validate the system wide metrics as we add and remove instances.
+			validateMetrics := func(expectedSilences int) {
+				// Check aggregate metrics across all instances.
+				ams := e2ecortex.NewCompositeCortexService(instances...)
+
+				// All instances should discover all tenants.
+				require.NoError(t, ams.WaitSumMetrics(
+					e2e.Equals(float64(numUsers*len(instances))),
+					"cortex_alertmanager_tenants_discovered"))
+
+				// If the number of instances has not yet reached the replication
+				// factor, then effective replication will be reduced.
+				var expectedReplication int
+				if len(instances) <= testCfg.replicationFactor {
+					expectedReplication = len(instances)
+				} else {
+					expectedReplication = testCfg.replicationFactor
+				}
+
+				require.NoError(t, ams.WaitSumMetrics(
+					e2e.Equals(float64(numUsers*expectedReplication)),
+					"cortex_alertmanager_tenants_owned"))
+
+				require.NoError(t, ams.WaitSumMetrics(
+					e2e.Equals(float64(numUsers*expectedReplication)),
+					"cortex_alertmanager_config_last_reload_successful"))
+
+				require.NoError(t, ams.WaitSumMetrics(
+					e2e.Equals(float64(expectedSilences*expectedReplication)),
+					"cortex_alertmanager_silences"))
+			}
+
+			// Start up the first instance and use it to create some silences.
+			numSilences := 0
+			{
+				am1 := startInstance()
+
+				// Validate metrics with only the first instance running, before creating silences.
+				validateMetrics(0)
+
+				// Only create silences for every other user. It will be common that some users
+				// have no silences, so some irregularity in the test is beneficial.
+				for i := 1; i <= numUsers; i += 2 {
+					user := fmt.Sprintf("user-%d", i)
+					client, err := e2ecortex.NewClient("", "", am1.HTTPEndpoint(), "", user)
+					require.NoError(t, err)
+
+					for j := 1; j <= 10; j++ {
+						silence := types.Silence{
+							Matchers: amlabels.Matchers{
+								{Name: "instance", Value: "prometheus-one"},
+							},
+							Comment:  fmt.Sprintf("Silence Comment #%d", j),
+							StartsAt: time.Now(),
+							EndsAt:   time.Now().Add(time.Hour),
+						}
+
+						_, err = client.CreateSilence(context.Background(), silence)
+						assert.NoError(t, err)
+						numSilences++
+					}
+				}
+
+				// Validate metrics after creating silences.
+				validateMetrics(numSilences)
+
+				// If we are testing with persistence, then check the persister actually activated.
+				// It's unlikely that nothing has been persisted by now if correctly configured.
+				if testCfg.withPersister {
+					require.NoError(t, instances[0].WaitSumMetrics(
```

Comment:
[nit] You can be even more explicit picking the AM instance.

```suggestion
					require.NoError(t, am1.WaitSumMetrics(
```